aws/src/main/java/org/apache/iceberg/aws/lakeformation/LakeFormationAwsClientFactory.java

```diff
@@ -132,27 +145,90 @@ private LakeFormationClient lakeFormation() {
         .build();
   }
 
+  static class LakeFormationTemporaryCredentials {
+
+    private final AwsCredentials credentials;
+    private final Instant sessionCredentialsExpiration;
+
+    LakeFormationTemporaryCredentials(AwsCredentials credentials, Instant expiration) {
+      this.credentials = credentials;
+      this.sessionCredentialsExpiration = expiration;
+    }
+
+    AwsCredentials credentials() {
+      return credentials;
+    }
+
+    Instant sessionCredentialsExpiration() {
+      return sessionCredentialsExpiration;
+    }
+  }
+
   static class LakeFormationCredentialsProvider implements AwsCredentialsProvider {
     private LakeFormationClient client;
     private String tableArn;
+    private final long expiryLeadTimeForRefreshInSecs;
+
+    private LakeFormationTemporaryCredentials cachedCredentials;
+    private final Cache<String, LakeFormationTemporaryCredentials> cache;
 
-    LakeFormationCredentialsProvider(LakeFormationClient lakeFormationClient, String tableArn) {
+    LakeFormationCredentialsProvider(LakeFormationClient lakeFormationClient,
+        String tableArn, long expiryLeadTimeForRefreshInSecs) {
       this.client = lakeFormationClient;
       this.tableArn = tableArn;
+      this.expiryLeadTimeForRefreshInSecs = expiryLeadTimeForRefreshInSecs;
+      this.cache = Caffeine.newBuilder()
+          .expireAfter(new Expiry<String, LakeFormationTemporaryCredentials>() {
+            @Override
+            public long expireAfterCreate(@NonNull String key,
```

Comment:
nit: if we cannot fit all arguments on the same line, let's put one argument per line, like:

```
public long expireAfterCreate(
  @NonNull String key,
  @NonNull LakeFormationTemporaryCredentials value, 
  long currentTime) {
  ...
}
```

Reply:
also I don't think we need these `@NonNull` and `@NonNegative` decorators?